### PR TITLE
feat(instance-search): create computed field for sorting and filtering Date1

### PR DESCRIPTION
**Pull request by SvitlanaKovalova1**
### Purpose
[MSEARCH-806](https://folio-org.atlassian.net/browse/MSEARCH-806) Currently in the Instance schema, there is a field for publication date. However, this data originates from an uncontrolled string in variable fields of the MARC record. The dates in the 008 are more consistently formatted, and are more useful for certain materials like serials. Previously, we implemented sortable properties (publicationPeriod.start/publicationPeriod.end) that are parsed from the publication date field. We need to break this relationship, and instead make the source of that data the 07-14 from the 008. The 008 structure allows for a single position to indicate the type of date that will be represented in the following positions.
008/06 = Date type
008/07-10 = Date 1
008/11-14 = Date 2

For this story, we need to create computed fields from Date 1 for sorting and filtering dates. And we need validation on the backend that stores those validation rules, and if the values in Date 1 don’t pass validation, we should consider as zeroes so that they still show in results list and show in filters

### Requirements:
- If Date 1 contains “u” in one of the four characters, replace “u” with 0
- If no data, consider as 0
- Store validation rules for Date 1:
      - Each can only store up to four characters
      - Allows for numeric values and the alpha “u”
- If characters in Date 1 do not pass this validation, consider the values as 0
- Return Date 1, Date 2, and Date type
- Verify the ability to search original data (not computed field) in “All” search; example queries:
      - Support trailing, leading, and internal wildcards and exact matches (15*, *02, 19*7, 1970)
      - Dates with “u” should follow the original data (not the computed field with “u” replaced with “0”): 199u exact match, would also match “199*”

### Approach
Update index:
1. Add new Instance schema fields: dateTypeId, date 1, date 2. It is searchable based on requirements.
2. Add a new searchable field: sort_date1.  It is computed from date 1 based on requirements. It  works with range queries and is used for sorting.

### Changes Checklist
- [ ] **API Changes**: Document any API paths, methods, request or response bodies changed, added, or removed.
- [ ] **Database Schema Changes**: Indicate any database schema changes and their impact. Confirm that migration scripts were created.
- [ ] **Interface Version Changes**: Indicate any changes to interface versions.
- [ ] **Interface Dependencies**: Document added or removed dependencies.
- [ ] **Permissions**: Document any changes to permissions.
- [ ] **Logging**: Confirm that logging is appropriately handled.
- [x] **Unit Testing**: Confirm that changed classes were covered by unit tests.
- [x] **Integration Testing**: Confirm that changed logic was covered by integration tests.
- [ ] **Manual Testing**: Confirm that changes were tested on local or dev environment.
- [ ] **NEWS**: Confirm that the NEWS file is updated with relevant information about the changes made in this pull request.

### CODE REVIEW

File: src/main/resources/model/instance.json [hunk 1]
```diff
@@ -127,6 +127,23 @@
...
+          "showInResponse": [ "search"]
+        },
+        "date1": {
+          "index": "keyword_lowercase",
+          "showInResponse": [ "search" ]
+        },
+        "date2": {
+          "index": "keyword_lowercase",
+          "showInResponse": [ "search" ]
+        }
```
Comment (psmagin) [inline src/main/resources/model/instance.json:144]: This 2 also should be `source` as we shouldn't support search directly by them.

File: README.md [hunk 2]
```diff
@@ -559,6 +559,9 @@ does not produce any values, so the following search options will return an empt
...
 | `normalizedClassificationNumber`       |   term    | `normalizedClassificationNumber == "LCCN"`                           | Matches instances with the given classification number (normalizes case, whitespaces, special characters, supports leading and trailing wildcard) |
+| `dates.date1`                          |   term    | `dates.date1="199*"`                                                 | Matches instances with the given Date1 (supports leading, trailing  and internal wildcards)                                                       |
+| `dates.date2`                          |   term    | `dates.date2="199*"`                                                 | Matches instances with the given Date2 (supports leading, trailing  and internal wildcards)                                                       |
+| `sort_date1`                           |   term    | `sort_date1=199u`                                                    | Matches instances with the given Date1 (normalizes  alpha 'u' characters, supports leading, trailing  and internal wildcards)                     |
```
Comment (psmagin) [inline README.md:564]: Why not `normalizedDate1` ?

Comment (psmagin) [reply README.md:564]:
```suggestion
| `sort_date1`                           |   term    | `sort_date1=1990`                                                    | Matches instances with the given Date1 (normalizes  alpha 'u' characters, supports leading, trailing  and internal wildcards)                     |
```
As you don't do input term normalization.

File: src/main/resources/model/instance.json [hunk 3]
```diff
@@ -552,6 +569,12 @@
       "index": "keyword_lowercase",
       "processor": "sortTitleProcessor"
     },
+    "sort_date1": {
+      "searchTypes": [ "filter", "sort" ],
+      "type": "search",
+      "index": "keyword_lowercase",
```
Comment (psmagin) [inline src/main/resources/model/instance.json:575]: I think we should make it an `number` type to make sure that we will have correct number-based order.

File: src/main/java/org/folio/search/service/setter/instance/Date1FieldProcessor.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.folio.search.domain.dto.Dates;
+import org.folio.search.domain.dto.Instance;
+import org.folio.search.service.setter.FieldProcessor;
+import org.springframework.stereotype.Component;
+
+@Component
+public class Date1FieldProcessor implements FieldProcessor<Instance, String> {
+
+  private static final Pattern FOUR_DIGIT_REGEX = Pattern.compile("^\\d{4}$");
```
Comment (psmagin) [inline src/main/java/org/folio/search/service/setter/instance/Date1FieldProcessor.java:13]: It could be less than 4 numbers, like 998, 503, 23, 1

**Review by psmagin — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (SvitlanaKovalova1) [reply README.md]: The "sort_" prefix in the field name is required for the sortBy operation. By default, the "sort_" prefix is ​​added during the sortBy operation in CqlSortProvider, and if it cannot find a field with this prefix, we get this error: "No mapping found for [sort_normalizedDate1] in order to sort on".

- field name: "normalizedDate1"
- error:
image

File: src/main/resources/model/instance.json (hunk 3 above)
Comment (SvitlanaKovalova1) [reply src/main/resources/model/instance.json]: updated

File: src/main/java/org/folio/search/service/setter/instance/Date1FieldProcessor.java (hunk 4 above)
Comment (SvitlanaKovalova1) [reply src/main/java/org/folio/search/service/setter/instance/Date1FieldProcessor.java]: updated

File: src/main/resources/model/instance.json (hunk 1 above)
Comment (SvitlanaKovalova1) [reply src/main/resources/model/instance.json]: updated

File: README.md (hunk 2 above)
Comment (psmagin) [reply README.md]: Could we do smth similar to authority mappings?
image

File: src/test/java/org/folio/search/service/setter/instance/Date1FieldProcessorTest.java [hunk 5]
```diff
@@ -20,21 +20,23 @@ class Date1FieldProcessorTest {
...
+      arguments(instance("20u2"), Short.valueOf("2002")),
+      arguments(instance("20uu"), Short.valueOf("2000")),
+      arguments(instance("199"), Short.valueOf("199")),
+      arguments(instance("1u9"), Short.valueOf("109")),
+      arguments(instance("1"), Short.valueOf("1")),
+      arguments(instance("19999"), Short.valueOf("0")),
+      arguments(instance("19k5"), Short.valueOf("0")),
+      arguments(new Instance(), Short.valueOf("0"))
     );
   }
```
Comment (psmagin) [inline src/test/java/org/folio/search/service/setter/instance/Date1FieldProcessorTest.java:41]: What about leading zeros? Cases: 0999, uuu9

File: README.md [hunk 6]
```diff
@@ -559,6 +559,9 @@ does not produce any values, so the following search options will return an empt
...
 | `normalizedClassificationNumber`       |   term    | `normalizedClassificationNumber == "LCCN"`                           | Matches instances with the given classification number (normalizes case, whitespaces, special characters, supports leading and trailing wildcard) |
+| `dates.date1`                          |   term    | `dates.date1="199*"`                                                 | Matches instances with the given Date1 (supports leading, trailing  and internal wildcards)                                                       |
+| `dates.date2`                          |   term    | `dates.date2="199*"`                                                 | Matches instances with the given Date2 (supports leading, trailing  and internal wildcards)                                                       |
+| `sort_date1`                           |   term    | `sort_date1=1990`                                                    | Matches instances with the given Date1 (normalizes  alpha 'u' characters, supports leading, trailing  and internal wildcards)                     |
```
Comment (psmagin) [inline README.md:564]: So it doesn't normalize input

File: src/test/java/org/folio/search/service/setter/instance/Date1FieldProcessorTest.java (hunk 5 above)
Comment (psmagin) [reply src/test/java/org/folio/search/service/setter/instance/Date1FieldProcessorTest.java:41]: Could you add it as a test case?

**Review by psmagin — APPROVED** [review]
Good job!

**Review by viacheslavkol — APPROVED** [review]

**Pull request merged**